File: solutions/java/src/main/java/com/epi/SquareRoot.java
package com.epi;

import java.util.Random;

public class SquareRoot {
  // @include
  public static double squareRoot(double x) {
    // Decides the search range according to x's value relative to 1.0.
    double left, right;
    if (x < 1.0) {
      left = x;
      right = 1.0;
    } else { // x >= 1.0.
      left = 1.0;
      right = x;
    }

    // Keeps searching as long as left < right, within tolerance.
    while (compare(left, right) == Ordering.SMALLER) {
      double mid = left + 0.5 * (right - left);
      double midSquared = mid * mid;
      if (compare(midSquared, x) == Ordering.EQUAL) {
        return mid;
      } else if (compare(midSquared, x) == Ordering.LARGER) {
        right = mid;
      } else {
        left = mid;
      }
    }
    return left;
  }

  private static enum Ordering { SMALLER, EQUAL, LARGER }

  private static Ordering compare(double a, double b) {
    final double EPSILON = 0.00001;
    // Uses normalization for precision problem.
    double diff = (a - b) / b;
    return diff < -EPSILON
        ? Ordering.SMALLER
        : (diff > EPSILON ? Ordering.LARGER : Ordering.EQUAL);
  }
  // @exclude

  private static void simpleTest() {
    Double[] res = new Double[2];
    res[0] = squareRoot(1.0);
    res[1] = Math.sqrt(1.0);
    assert(compare(res[0], res[1]) == Ordering.EQUAL);

    res[0] = squareRoot(2.0);
    res[1] = Math.sqrt(2.0);
    assert(compare(res[0], res[1]) == Ordering.EQUAL);

    res[0] = squareRoot(0.001);
    res[1] = Math.sqrt(0.001);
    assert(compare(res[0], res[1]) == Ordering.EQUAL);

    res[0] = squareRoot(0.5);
    res[1] = Math.sqrt(0.5);
    assert(compare(res[0], res[1]) == Ordering.EQUAL);

    res[0] = squareRoot(10000000000.001);
    res[1] = Math.sqrt(10000000000.001);
    assert(compare(res[0], res[1]) == Ordering.EQUAL);

    res[0] = squareRoot(1024.0);
    res[1] = Math.sqrt(1024.0);
    assert(compare(res[0], res[1]) == Ordering.EQUAL);
  }

  public static void main(String[] args) {
    simpleTest();
    Random r = new Random();
    for (int times = 0; times < 100000; ++times) {
      double x;
      if (args.length == 1) {
        x = Double.valueOf(args[0]);
      } else {
        x = r.nextDouble() * 100000000;
      }
      double[] res = new double[2];
      System.out.println("x is " + x);
      System.out.println((res[0] = squareRoot(x)) + " "
                         + (res[1] = Math.sqrt(x)));
      assert(compare(res[0], res[1]) == Ordering.EQUAL);
    }
  }
}

File: static/solutions/Square_root.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <cmath>
#include <iostream>
#include <limits>
#include <random>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::numeric_limits;
using std::random_device;
using std::uniform_real_distribution;

// @include
typedef enum { SMALLER, EQUAL, LARGER } Ordering;

// @exclude
Ordering Compare(double a, double b);
// @include
double SquareRoot(double x) {
  // Decides the search range according to x's value relative to 1.0.
  double left, right;
  if (x < 1.0) {
    left = x, right = 1.0;
  } else {  // x >= 1.0.
    left = 1.0, right = x;
  }

  // Keeps searching as long as left < right, within tolerance.
  while (Compare(left, right) == SMALLER) {
    double mid = left + 0.5 * (right - left);
    double mid_squared = mid * mid;
    if (Compare(mid_squared, x) == EQUAL) {
      return mid;
    } else if (Compare(mid_squared, x) == LARGER) {
      right = mid;
    } else {
      left = mid;
    }
  }
  return left;
}

Ordering Compare(double a, double b) {
  // Uses normalization for precision problem.
  double diff = (a - b) / b;
  return diff < -numeric_limits<double>::epsilon()
             ? SMALLER
             : diff > numeric_limits<double>::epsilon() ? LARGER : EQUAL;
}
// @exclude

static void SimpleTest() {
  assert(Compare(SquareRoot(1.0), sqrt(1.0) == EQUAL));
  assert(Compare(SquareRoot(2.0), sqrt(2.0) == EQUAL));
  assert(Compare(SquareRoot(0.001), sqrt(0.001) == EQUAL));
  assert(Compare(SquareRoot(0.5), sqrt(0.5) == EQUAL));
  assert(Compare(SquareRoot(100000000.001), sqrt(100000000.001) == EQUAL));
  assert(Compare(SquareRoot(1024.0), sqrt(1024.0) == EQUAL));
}

int main(int argc, char* argv[]) {
  SimpleTest();
  default_random_engine gen((random_device())());
  for (int times = 0; times < 100000; ++times) {
    double x;
    if (argc == 2) {
      x = atof(argv[1]);
    } else {
      uniform_real_distribution<double> dis(0.0, 100000000.0);
      x = dis(gen);
    }
    double res[2];
    cout << "x is " << x << endl;
    cout << (res[0] = SquareRoot(x)) << ' ' << (res[1] = sqrt(x)) << endl;
    assert(Compare(res[0], res[1]) == EQUAL);
  }
  return 0;
}
